### DISCUSSION TITLE
Is it possible to deploy patched node modules?

### DISCUSSION THREAD

**Discussion by snelgrove** [architect/architect#1265] [Q&A]
I've been trying to deploy a patched third-party node module in a http route, but the original pre-patched version is deployed instead. 

I made a small change to a module, and saved the patch using https://github.com/ds300/patch-package. This seems to be a really neat way to make a minor change to a module. After adding the `"postinstall": "patch-package"` NPM script, my patch gets applied any time `npm i` is called, and I can see my change in the project level `node_modules` directory.

My http function does not have a package.json, so I'm relying on `arc` to work out the dependencies. When I `arc deploy`, however, my function gets the original module, pre-patch. I had assumed the project level modules would get deployed, but it looks like these modules are cached elsewhere. 

Despite searching the various architect repositories, I couldn't work out how/why this was happening. Any suggestions on how to make this work? 

Thanks!

**Comment by tbeseda** [comment]
Ah yeah `patch-package` has some pretty cool uses! I like that it uses .patch files.

Anyway, I'm not totally sure why the deploy with Arc wouldn't run the patch, but it might be possible to move and patch as part of the deploy. We do something similar in the arc.codes project to reduce the size of the highlight.js dependency https://github.com/architect/arc.codes/blob/main/scripts/dist.sh  
Basically move select files from the node_modules folder directly to the function source and require it from there.

There's also a call to that script on sandbox startup so it works locally https://github.com/architect/arc.codes/blob/main/prefs.arc

**Reply by snelgrove** [reply]
Thanks for your advice. I like this approach, but it my case it started getting a bit complicated because I'm using TypeScript and the NPM module is JS. I should have also mentioned I'm also using Architect's custom paths for separate src and build directories. 

It still seems that somewhere in the build/sandbox/deploy, the original NPM Module is included rather than my local patched copy. I'm sure it's because I've gone rogue by adopting TS, custom paths and now patched 3rd party modules 😅 

Anyway, after trying many different angles, I have found it's just easier and tidy enough to fork the repo, create a patched branch and install it straight from github using `npm i <user>/<repo>#<branch>`.

Thanks again!

**Reply by ryanblock** [reply]
What's happening under the hood: `npm [c]i` (or `yarn install`) is being run at the very last moment before deployment.

Your fork and install seems like a solid workaround. Another approach might to do all your own dependency management prior to your deploy step, then `[npx] arc deploy --no-hydrate` (which does what it sounds like it does).

I'm open to other ideas for the Arc side too, if you can think of any!

**Reply by snelgrove** [reply]
My understanding is that `npm ci` relies on `package-lock.json` and installs the exact versions referenced there. I'm not sure where it sources them from though - I guess it's not the project's node_modules directory. 

Having Arc apply the patches upon hydrate seems way too specific, but maybe there's a way Arc could (optionally) hydrate from `<project-root>/node_modules`?

**Reply by ryanblock** [reply]
Hmm, sorry, no way to do that, npm is responsible for resolving the dependency tree and you don’t want the entire project’s node modules dir in every lambda, just the tree-shaken subset required to run that particular function.
